crates/red_knot_python_semantic/resources/mdtest/doc/public_type_undeclared_symbols.md

```diff
@@ -0,0 +1,115 @@
+# Public type of undeclared symbols
+
+## Summary
+
+One major deviation from the behavior of existing Python type checkers is our handling of 'public'
+types for undeclared symbols. This is best illustrated with an example:
+
+```py
+class Wrapper:
+    value = None
+
+wrapper = Wrapper()
+
+reveal_type(wrapper.value)  # revealed: Unknown | None
+
+wrapper.value = 1
+```
+
+Mypy and Pyright both infer a type of `None` for the type of `wrapper.value`. Consequently, both
+tools raise an error when trying to assign `1` to `wrapper.value`. But there is nothing wrong with
+this program. Raising an error here violates the [gradual guarantee] which states that *"Removing
+type annotations (making the program more dynamic) should not result in additional static type
+errors."*: If `value` were annotated with `int | None` here, Mypy and Pyright would not raise any
+errors.
+
+By inferring `Unknown | None` instead, we allow arbitrary values to be assigned to `wrapper.value`.
+This is a deliberate choice to prevent false positive errors on untyped code.
+
+More generally, we infer `Unknown | T_inferred` for undeclared symbols, where `T_inferred` is the
+inferred type of the right-hand side of the assignment. This gradual type represents an *unknown*
+fully-static type that is *at least as large as* `T_inferred`. It accurately describes our static
+knowlege about this type. In the example above, we don't know what values `wrapper.value` could
+possibly contain, but we *do know* that `None` is a possibility. This allows us to catch errors
+where `wrapper.value` is used in a way that is incompatible with `None`:
+
+```py
+def f(w: Wrapper) -> None:
+    # This is fine
+    x: int | None = w.value
+
+    # This is not okay because `w.value` could be `None`. We therefore raise the following
+    # error: "Object of type `Unknown | None` is not assignable to `int`"
+    y: int = w.value
```

Comment:
You could also use this as an additional example, which might be slightly more "realistic" since users often leave local variables unannotated:

```suggestion
    y: int = w.value

    z = w.value + 42  # error
```

Reply:
That's similar to what I wanted to demonstrate here at first, but unfortunately, we don't detect that `None + 42` is an error, yet.

Reply:
ah, we should probably fix that 🙃

Reply:
I really wanted this document to be user-facing, so I would prefer if it were free of TODOs. I now used a `chr(w.value)` call to demonstrate that we would emit a *"`Unknown | None` cannot be assigned to parameter 1 (`i`) of function `chr`; expected type `int`"* diagnostic.

I used `chr` (as opposed to something more prominent) because it is one of very few functions in `builtins` that "just takes an `int`" as a parameter, and not some complicated generic protocol :upside_down_face:

Reply:
> I used `chr` (as opposed to something more prominent) because it is one of very few functions in `builtins` that "just takes an `int`" as a parameter, and not some complicated generic protocol 🙃

Oh no, that... Seems like it might be a typeshed bug 😆

```pycon
>>> class NotAnInt:
...   def __index__(self): return 42
...   
>>> chr(NotAnInt())
'*'
```

I can keep quiet about it until red-knot's implemented generics and protocols, though 😉

Reply:
Couldn't you just as well define your own function that "just takes an int" instead of needing to use something from builtins?

Reply:
> Couldn't you just as well define your own function that "just takes an int"

Yes, I did that now. I wanted to keep it short, but I shouldn't trade brevity for clarity.

> Seems like it might be a typeshed bug 😆

Oh :smile:

https://github.com/python/typeshed/pull/13494